remotes/docker/pusher.go

```diff
@@ -44,17 +44,46 @@ type dockerPusher struct {
 	tracker StatusTracker
 }
 
+// Writer implements Ingester API of content store. This allows the client
+// to receive ErrUnavailable when there is already an on-going upload.
+// Note that the tracker MUST implement StatusTrackLocker interface to avoid
+// race condition on StatusTracker.
+func (p dockerPusher) Writer(ctx context.Context, opts ...content.WriterOpt) (content.Writer, error) {
+	var wOpts content.WriterOpts
+	for _, opt := range opts {
+		if err := opt(&wOpts); err != nil {
+			return nil, err
+		}
+	}
+	if wOpts.Ref == "" {
+		return nil, errors.Wrap(errdefs.ErrInvalidArgument, "ref must not be empty")
+	}
+
+	if l, ok := p.tracker.(StatusTrackLocker); ok {
+		l.Lock(wOpts.Ref)
+		defer l.Unlock(wOpts.Ref)
+	}
```

Comment:
Can this happen inside `push`?  The `ref` argument to that function can be used as the lock name since it's constructed as `remotes.MakeRefKey(ctx, desc)` in both `Push` and `remotes.push`.

Reply:
Fixed this.